code/ingest/src/main/java/org/jesterj/ingest/scanners/JdbcScanner.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2016 Needham Software LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jesterj.ingest.scanners;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jesterj.ingest.model.Document;
+import org.jesterj.ingest.model.Router;
+import org.jesterj.ingest.model.exception.ConfigurationException;
+import org.jesterj.ingest.model.exception.PersistenceException;
+import org.jesterj.ingest.model.impl.DocumentImpl;
+import org.jesterj.ingest.model.impl.ScannerImpl;
+import org.jesterj.ingest.utils.SqlUtils;
+
+import net.jini.space.JavaSpace;
+
+/**
+ * Scans a JDBC source such as an RDBMS (e.g. MySQL). Obtains a connection through the specified
+ * JDBC driver and extracts rows using the specified SQL statement; converts extracted rows to
+ * documents and passes the documents downstream.
+ * 
+ * @author dgoldenberg
+ */
+public class JdbcScanner extends ScannerImpl {
+
+  private static final Logger log = LogManager.getLogger();
+
+  private String jdbcDriver;
+  private String jdbcUrl;
+  private String jdbcUser;
+  private String jdbcPassword;
+
+  private String sqlStatement;
+
+  // TODO DG discuss
+  // May need to be set to Integer.MIN_VALUE for MySQL
+  // http://stackoverflow.com/questions/2447324/streaming-large-result-sets-with-mysql
+  // http://stackoverflow.com/questions/24098247/jdbc-select-batching-fetch-size-with-mysql
+  // http://stackoverflow.com/questions/20899977/what-and-when-should-i-specify-setfetchsize
+  private int fetchSize = -1;
+
+  private boolean autoCommit;
+  private int queryTimeout = -1;
+
+  @Override
+  public Function<String, String> getIdFunction() {
+    // TODO DG
+    return s -> s;
+  }
+
+  @Override
+  public Consumer<Document> getDocumentTracker() {
+    // TODO DG
+    return document -> {
+    };
+  }
+
+  @Override
+  public Runnable getScanOperation() {
+    return () -> {
```

Comment:
This pattern (which I started and you are quite reasonably following) may lead us into trouble with unit testing... http://stackoverflow.com/questions/28688047/unit-test-code-with-java-8-lambdas